databroker/eventsource/shim.py

```diff
@@ -5,18 +5,52 @@
 import pandas as pd
 import logging
 import boltons.cacheutils
+import heapq
 import re
 from ..utils import ALL
+
 # Toolz and CyToolz have identical APIs -- same test suite, docstrings.
 try:
     from cytoolz.dicttoolz import merge
 except ImportError:
     from toolz.dicttoolz import merge
 
-
 logger = logging.getLogger(__name__)
 
 
+def interlace_gens(*gens):
+    """Take generators and interlace their results by timestamp
+
+    Parameters
+    ----------
+    gens : generators
+        Generators of (name, dict) pairs where the dict contains a 'time'
+        key.
+
+    Yields
+    -------
+    val : tuple
+        The next (name, dict) pair in time order
+
+    """
+    iters = [iter(g) for g in gens]
+    heap = []
+
+    def safe_next(indx):
+        try:
+            val = next(iters[indx])
```

Comment:
could this be `val = next(gen)`?  I guess you are using the index to break the tie when two things have _exactly_ the same time?

Reply:
What is `gen`?

Reply:
`gen` should be `iter_gen` (then things you keep in `iters`).

Reply:
the `indx` is in there so we don't need to store the entire generators inside the heap. The `indx` is also returned in the `heappop`. This allows us to draw from a generator until it finally falls behind another entry.

Reply:
Isn't it an object either way?

```python
heap = []
def safe_next(itr, indx):
    try:
        val = next(itr)
    except StopIteration:
        return
    heapq.heappush(heap, (val['time'], indx, val, itr))

for j, itr in enumerate(iters):
    safe_next(itr, j)
while heap:
    _, j, val, itr = heapq.heappop(heap)
   yield val
   safe_next(itr, j)

```

would also work.

Reply:
where here the index is serving as the tie-breaker.